core/modules/loader/CentralClient.cc

```diff
@@ -0,0 +1,276 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "CentralClient.h"
+
+// system headers
+#include <iostream>
+
+// Third-party headers
+#include <boost/asio.hpp>
+
+// qserv headers
+#include "loader/LoaderMsg.h"
+#include "proto/loader.pb.h"
+#include "proto/ProtoImporter.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.loader.CentralClient");
+}
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+
+void CentralClient::handleKeyInfo(LoaderMsg const& inMsg, BufferUdp::Ptr const& data) {
+    LOGS(_log, LOG_LVL_DEBUG, "CentralClient::handleKeyInfo");
+
+    StringElement::Ptr sData = std::dynamic_pointer_cast<StringElement>(MsgElement::retrieve(*data));
```

Comment:
The buffer BufferUdp object passed to retrieve is often not associated with a shared pointer.
Also, I don't see why this would be unsafe as the data shared pointer object must exist until the end of the function.